### PR TITLE
OpenVINO-TorchORT Integration

### PR DESCRIPTION
**Pull request by saipj**
Description:

This PR enables inference in ORTModule to leverage existing OnnxRuntime OpenVINO EP. 

- Changes specific to inference are added in export parameters without any gradient components and custom symbolic changes are disabled.  
- ProviderConfigs is introduced to enable generic approach when using providers (OpenVINO EP here). 

Motivation:

- Supports export of models with Aten Fallback export option.
- Supports inference on multiple Intel backends such as CPUs, iGPUs and VPUs(MYRIADX).

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 1]
```diff
@@ -394,36 +407,65 @@ def _get_exported_model(self, input_schema, *inputs, **kwargs):
...
-                                  **self._export_extra_kwargs)
-        except Exception as e:
-            raise wrap_exception(ORTModuleONNXModelException,
-                                 RuntimeError(f'There was an error while exporting the PyTorch model to ONNX: '
-                                              f'\n\n{_utils.get_exception_as_string(e)}'))
-        exported_model = onnx.load_model_from_string(f.getvalue())
-
-        exported_model = _post_process_after_export(exported_model,
+        provider = self._provider_configs.provider
+        if not provider == "openvino":
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:411]: Could you clarify why we need to branch this part?

**Review by thiagocrepaldi — COMMENTED** [review]
A high-level explanation on the changes and why they were needed would be appreciated

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 1 above)
Comment (saipj) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:411]: @thiagocrepaldi, we are branching out to enable changes in export params specific to inference using openvino and avoid changes to existing path specific to ORT training that includes gradient related code.

**Comment by saipj** [comment]
> A high-level explanation on the changes and why they were needed would be appreciated

@thiagocrepaldi , updated the description section to provide high-level details.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 2]
```diff
@@ -71,13 +73,14 @@ def is_disabled(self):
 
 
 class GraphExecutionManager(GraphExecutionInterface):
-    def __init__(self, module, debug_options: DebugOptions, fallback_manager: _FallbackManager):
+    def __init__(self, module, debug_options: DebugOptions, fallback_manager: _FallbackManager, provider_configs = ProviderConfigs):
```
Comment (askhade) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:76]: What are the current blockers in using ortdevice? Per my knowledge all the build failures were resolved right?

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 1 above)
Comment (askhade) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:411]: The problem with this approach is that user does not know the implications of choosing openvino execution provider. When user chooses openvino, internally we run inference on the model which is very different than the usual flow

**Comment by askhade** [comment]
@saipj : Can you add some example usage for this mode.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 2 above)
Comment (saipj) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:76]: @askhade, there are multiple blockers - graph mode not supported for ORT device, data copy and transfer issues, and unsupported ops with 'ORT' backend.

**Comment by askhade** [comment]
Please add tests and make sure they include something to test aten fallback option as well

**Comment by askhade** [comment]
I think it would be better to add a flag for explicitly stating the mode as inference. Right now it is not easy for the user to understand that OpenVINO will only be available for inference.

**Comment by saipj** [comment]
> @saipj : Can you add some example usage for this mode.

Example usage in a sample:
from torch_ort import ORTModule, ProviderConfigs
provider_configs = ProviderConfigs(provider="openvino", backend="CPU",precision="FP32")
model = ORTModule(model, provider_configs=provider_configs)

**Comment by askhade** [comment]
> > @saipj : Can you add some example usage for this mode.
> 
> Example usage in a sample: from torch_ort import ORTModule, ProviderConfigs provider_configs = ProviderConfigs(provider="openvino", backend="CPU",precision="FP32") model = ORTModule(model, provider_configs=provider_configs)

Can you add the rest of the steps which follow this as well...

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 1 above)
Comment (souptc) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:411]: may i understand what kind of exporter params specific you need? is it because of openvino or is it because you only run inference? i am feeling we are mixed those two concepts. This OpenVino PR actually introduce two things: 1) add openvino provider option. 2) implicitly add a "inference mode" in ortmodule, for example, export all the weights as initializers, not inputs. 

If we really want to do it, we should seperate them into two tasks. one is pure enable openvino provider option. another is to provide "inference mode". The "inference mode" could be used for not only openvino, but also ort CPU/Cuda/TensorRT EPs. and we should have explicit check to avoid pytorch user accidently trigger training feature on "inference mode"

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 3]
```diff
@@ -394,36 +407,65 @@ def _get_exported_model(self, input_schema, *inputs, **kwargs):
...
+                   required_export_kwargs = {'input_names': self._input_info.names,
+                                             'output_names': output_names,
+                                             'opset_version': ortmodule.ONNX_OPSET_VERSION,
+                                             'do_constant_folding': True,
+                                             'training': self._export_mode,
+                                             'dynamic_axes': self._input_info.dynamic_axes,
+                                             'verbose': self._debug_options.logging.log_level < LogLevel.WARNING,
+                                             'operator_export_type': OperatorExportTypes.ONNX_ATEN_FALLBACK,
+                                             'export_params': True,
+                                             'keep_initializers_as_inputs': False}
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:454]: Doing this implies that a lot of the `torch.nn.Module` functionality is no longer usable. For example, users cannot query `model.named_parameters()` since they will be different from the initializers being used (inside the exported onnx model).
This might be ok for the inference scenario (because the parameters are not going to change in inference). But the branching in this code here is happening on the provider, not on the training mode.

File: orttraining/orttraining/python/training/ortmodule/ortmodule.py [hunk 4]
```diff
@@ -36,7 +36,7 @@ class ORTModule(torch.nn.Module):
         debug_options (:obj:`DebugOptions`, optional): debugging options for ORTModule.
     """
 
-    def __init__(self, module, debug_options=None):
+    def __init__(self, module, debug_options=None, provider_configs=None):
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/ortmodule.py:39]: I prefer not to expose these options through the constructor because this config is being used so as to set the execution provider (pytorch device type) in addition to setting the provider config.
So far, we have been using `torch.device` to select the execution provider. But after this change, it becomes ambiguous to the user whether to use `model.to(device)` or to use `model = ORTModule(model, provider_config)`. This might not be the best API from a user perspective. Do we have any other ways to select the provider (maybe use the environment variables so it is hidden)?

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 5]
```diff
@@ -279,22 +282,32 @@ def _get_session_config(self):
...
             if self._use_external_gpu_allocator:
                 provider_option_map["gpu_external_alloc"] = str(self._torch_alloc)
                 provider_option_map["gpu_external_free"] = str(self._torch_free)
                 provider_option_map["gpu_external_empty_cache"] = str(self._torch_empty_cache)
-            provider_options = [provider_option_map, {}]
+                provider_options = [provider_option_map, {}]
         elif self._device.type == 'cpu':
-            providers = ["CPUExecutionProvider"]
-            provider_options = [{}]
+            if self._provider_configs.provider == "openvino":
```
Comment (souptc) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:299]: i feel we can make it more general. the real thing we need here is to have a mapping from pytorch device to available OR EPs. a single pytorch device may map to multiple EPs, with priority, i.e. CPU device map to OpenVINO EP and CPU EP, cuda device map to TensorRT EP and cuda EP.  So instead of have a configuration on ORTModule for EPs, i prefer to have an option let user configure the available EP priorities, then ORTModule can figure out the correct mapping based on that. something like:

import torch
import onnxruntime

onnxruntime.set_ep_priority('OpenVINOEP', 'CPUEP')
...
model = ORTModule(model)
...

with the priority configuration, we can have a generic algo here to generate the correct mapping, without change ortmodule api.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 1 above)
Comment (askhade) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:411]: "we should have explicit check to avoid pytorch user accidently trigger training feature on "inference mode" very true!

Comment (saipj) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:411]: Export params specific to inference that are needed:
'do_constant_folding': True,
'operator_export_type': OperatorExportTypes.ONNX_ATEN_FALLBACK,
'export_params': True,
'keep_initializers_as_inputs': False

Since these params are not specific to openvino, we can have "inference mode" for more generic usage across other ort EPs and avoid accidental trigger by pytorch user. We will update the PR shortly for these changes.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 6]
```diff
@@ -394,36 +407,65 @@ def _get_exported_model(self, input_schema, *inputs, **kwargs):
...
+                                    RuntimeError(f'There was an error while exporting the PyTorch model to ONNX: '
+                                                 f'\n\n{_utils.get_exception_as_string(e)}'))
+           exported_model = onnx.load_model_from_string(f.getvalue())
+
+           exported_model = _post_process_after_export(exported_model,
                                                     self._enable_custom_autograd_function,
                                                     self._debug_options.logging.log_level)
+        else:
+           try:
+               with torch.no_grad():
```
[the remark this answers is not in the document]
Comment (saipj) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:444]: 1. In "inference mode", we will use torch.no_grad() to disable gradient calculations and reduce memory usage for inference to get optimal performance. 
2. Currently openvino is used for inference only.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (saipj) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:299]: Noted. We will add the checks for openvino availability and update the PR.

**Comment by snnn** [comment]
This pull request has been automatically closed because it has merge conflicts and has been inactive for more than 30 days. Please rebase on the target branch and open a new PR.
